Mathlib/Tactic/ProdAssoc.lean

```diff
@@ -0,0 +1,150 @@
+      return fst' ++ snd'
+
+/-- This function should act as the "reverse" of `ProdTree.unpack`, constructing
+a term of the iterated product out of a list of terms of the types appearing in the product. -/
+def ProdTree.pack (ts : List Expr) : ProdTree → MetaM Expr
+  | type _ _ => do
+    match ts with
+      | [] => throwError "Can't pack the empty list."
+      | [a] => return a
+      | _ => throwError "Failed due to size mismatch."
+  | prod fst snd u v => do
+    let fstSize := fst.size
+    let sndSize := snd.size
+    unless ts.length == fstSize + sndSize do throwError "Failed due to size mismatch."
+    let tsfst := ts.toArray[:fstSize] |>.toArray.toList
+    let tssnd := ts.toArray[fstSize:] |>.toArray.toList
+    let mk : Expr := mkAppN (.const ``Prod.mk [u,v]) #[fst.getType, snd.getType]
+    return .app (.app mk (← fst.pack tsfst)) (← snd.pack tssnd)
+
+/-- Converts a term `e` in an iterated product `P1` into a term of an iterated product `P2`.
+Here `e` is an `Expr` representing the term, and the iterated products are represented
+by terms of `ProdTree`. -/
+def ProdTree.convertTo (P1 P2 : ProdTree) (e : Expr) : MetaM Expr :=
+  return ← P2.pack <| ← P1.unpack e
+
+/-- Given two expressions corresponding to iterated products of the same types, associated in
+possibly different ways, this constructs the "obvious" function from one to the other. -/
+def mkProdFun (a b : Expr) : MetaM Expr := do
+  let pa ← a.mkProdTree
+  let pb ← b.mkProdTree
+  unless pa.components.length == pb.components.length do
+    throwError "The number of components of {a} and {b} must match."
+  for (x,y) in pa.components.zip pb.components do
+    unless ← isDefEq x y do
+      throwError "Components {x} is not defeq to component {y}."
+  return .lam `t a (← pa.convertTo pb <| .bvar 0) .default
+
+/-- Construct the equivalence between iterated products of the same type, associated
+in possibly different ways. -/
+def mkProdEquiv (a b : Expr) : MetaM Expr := do
+  let some u := (← inferType a).type? | throwError "Not a type{indentExpr a}"
+  let some v := (← inferType b).type? | throwError "Not a type{indentExpr b}"
+  return mkAppN (.const ``Equiv.mk [.succ u,.succ v])
+    #[a, b, ← mkProdFun a b, ← mkProdFun b a,
+      .app (.const ``rfl [.succ u]) a,
+      .app (.const ``rfl [.succ v]) b]
+
+/-- IMPLEMENTATION: Syntax used in the implementation of `prod_assoc%`. -/
+syntax (name := prodAssocStx) "prod_assoc_internal%" : term
+
+open Elab Term in
+/-- Elaborator for `prod_assoc%`. -/
+@[term_elab prodAssocStx]
+def elabProdAssoc : TermElab := fun stx expectedType? => do
+  match stx with
+  | `(prod_assoc_internal%) => do
+    let some expectedType ← tryPostponeIfHasMVars? expectedType?
+          | throwError "expected type must be known"
+    let .app (.app (.const ``Equiv _) a) b := expectedType
+          | throwError "Expected type {expectedType} is not of the form `α ≃ β`."
```

Comment:
This error should never happen, but let's make it look nice anyway.
```suggestion
          | throwError "Expected type{indentD expectedType}\nis not of the form `α ≃ β`."
```

Reply:
Thanks. Done!